### ISSUE TITLE
--enable_bzlmod breaks toolchain resolution for toolchains registered in WORKSPACE

### ISSUE THREAD

**Issue by gregmagolan** [bazelbuild/bazel#17289]
### Description of the bug:

Setting the `--enable_bzlmod` flag (even with an empty `MODULE.bazel` file) breaks toolchain resolution for toolchains registered in the WORKSPACE file.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

https://github.com/gregmagolan/repro-enable_bzlmod_breaks_workspace

### Which operating system are you running Bazel on?

MacOS

### What is the output of `bazel info release`?

```
release 6.0.0
```

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

https://github.com/gregmagolan/repro-enable_bzlmod_breaks_workspace

### Have you found anything relevant by searching the web?

Searched through Bazel issues and didn't see any duplicates.

**Comment by katre** [comment]
This feels to be Working As Intended, since modules specifically have a different way to register toolchains, but I'll leave it for @Wyverald to clarify.

**Comment by gregmagolan** [comment]
I passed this by @Wyverald on Bazel Slack and he asked to create an issue and cc him. Was just trying to find with GitHub handle :)

If WORKSPACE files can't register toolchains then it will be hard to use rulesets that have bzlmod support alongside rulesets that don't (while they all get up to speed) if they both register toolchains.

**Comment by brentleyjones** [comment]
Not stale.

**Comment by Wyverald** [comment]
Ah, I think I figured out what's wrong here. It's that damn `platforms` repo again...

When you don't enable Bzlmod, `@platforms` for everyone is the `@@platforms` repo defined in a WORKSPACE prefix/suffix (don't remember which). All good.

When you enable Bzlmod, `@platforms` is still the `@@platforms` repo defined in a WORKSPACE prefix/suffix for everything that is *not* in the Bzlmod system (i.e. all the WORKSPACE repos including `@@rules_nodejs`). However, for things in the Bzlmod system (i.e. only built-in modules like `bazel_tools` and `local_config_platform` because your MODULE.bazel file is empty), `@platforms` actually refers to `@@platforms~0.0.5`. This is important because when we search for an appropriate execution platform (most likely your host machine), `@local_config_platform//:host` actually uses the platforms defined in `@@platforms~0.0.5`, but `rules_nodejs` defines toolchains for platforms defined in `@@platforms`.

The immediate workaround would be to add a single line in your MODULE.bazel file (`bazel_dep(name="platforms", version="0.0.5")`). I'll have to ponder how we could resolve this situation better, if it's even possible.

**Comment by HappyCerberus** [comment]
> Closing this as things are technically working as intended here. Hopefully the existence of this issue will serve as documentation for whoever's encountering this problem in the future.

Uhm, the module system not resolving a dependency, and the user having to manually depend on a specific version of a deep dependency is "working as intended"?

In my view this completely invalidates bzlmod.

It might be working "as designed", but definitely not "as intended".

**Comment by mtribiere** [comment]
Agreed, letting the user add a dependency to make it work isn't really a solution to me.
To anyone that helps: 
Add to `MODULE.bazel`:
```
bazel_dep(name="platforms", version="0.0.5")
```

**Comment by Wyverald** [comment]
if anyone has an idea how to fix this, please leave a suggestion.

**Comment by mtribiere** [comment]
Making making `platforms` v0.0.5 the default on Bazel 7 would be a solution ?
